kotlinx-coroutines-core/common/src/internal/SegmentQueueSynchronizer.kt

```diff
@@ -0,0 +1,606 @@
+ * This way, in [SMART_ASYNC] mode the value passed to [resume] can be out of the data structure for a while.
+ *
+ * Here is a state machine for cells. Note that only one [suspend]
+ * and at most one [resume] operation can deal with each cell.
+ *
+ *  +-------+   `suspend` succeeds.   +--------------+  `resume` is   +---------------+  store `RESUMED` to   +---------+  ( `cont` HAS BEEN   )
+ *  |  NULL | ----------------------> | cont: active | -------------> | cont: resumed | --------------------> | RESUMED |  ( RESUMED AND THIS  )
+ *  +-------+                         +--------------+  successful.   +---------------+  avoid memory leaks.  +---------+  ( `resume` SUCCEEDS )
+ *     |                                     |
+ *     |                                     | The continuation
+ *     | `resume` comes to                   | is cancelled.
+ *     | the cell before                     |
+ *     | `suspend` and puts                  V                                                                      ( THE CORRESPONDING `resume` SHOULD BE   )
+ *     | the element into           +-----------------+    The concurrent `resume` should be refused,    +--------+ ( REFUSED AND `tryReturnRefusedValue` OR )
+ *     | the cell, waiting          | cont: cancelled | -----------------------------------------------> | REFUSE | ( `returnValue` IF IT FAILS IS USED TO   )
+ *     | for `suspend` in           +-----------------+      `onCancellation` has returned `false`.      +--------+ ( RETURN THE VALUE BACK TO THE ORIGINAL  )
+ *     | SYNC mode.                          |         \                                                     ^      ( SYNCHRONIZATION PRIMITIVE              )
+ *     |                                     |          \                                                    |
+ *     |        Mark the cell as `CANCELLED` |           \                                                   |
+ *     |         if the cancellation mode is |            \  `resume` delegates its completing to            | `onCancellation` returned `false,
+ *     |        `SIMPLE` or `onCancellation` |             \   the concurrent cancellation handler if        | mark the state accordingly and
+ *     |                has returned `true`. |              \   `SMART_ASYNC` cancellation mode is used.     | complete the hung `resume`.
+ *     |                                     |               +------------------------------------------+    |
+ *     |                                     |                                                           \   |
+ *     |    ( THE CONTINUATION IS    )       V                                                            V  |
+ *     |    ( CANCELLED AND `resume` ) +-----------+                                                     +-------+
+ *     |    ( EITHER FAILS IN THE    ) | CANCELLED | <-------------------------------------------------- | value |
```

Comment:
I think that "EITHER" implies that any of the two outcomes is likely, but here we know in advance which one is the one. Removing the word "EITHER" fixes this.